common/sealing/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+//! [AEAD][aead].
+//!
+//! [aead]: https://en.wikipedia.org/wiki/Authenticated_encryption
+//! [chacha]: https://en.wikipedia.org/wiki/ChaCha20-Poly1305
+#![no_std]
+
+extern crate alloc;
+
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+
+use ring_compat::aead::{AeadInPlace, ChaCha20Poly1305, KeyInit};
+use ring_compat::generic_array::typenum::{U12, U32};
+use ring_compat::generic_array::GenericArray;
+pub use zeroize::Zeroizing;
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// A nonce should only ever be used once for any given key, across your entire
+/// code base.  Nonces may either be randomly generated or, otherwise, carefully
+/// selected to ensure uniqueness.
+#[derive(Debug)] // core
+#[derive(Zeroize, ZeroizeOnDrop)] // zeroize
+pub struct Nonce([u8; 12]);
+
+impl Nonce {
+    // The following constant is fixed by the ChaCha20-Poly1305 algorithm.
+    //
+    // TODO: figure out how to use the associated [`NonceSize`] type here
+    // instead.
+    pub const SIZE: usize = 12;
+
+    pub fn new(nonce: [u8; Self::SIZE]) -> Self {
+        Nonce::from(nonce)
+    }
+}
+
+#[derive(Debug, Clone)] // core
+#[derive(Zeroize, ZeroizeOnDrop)] // zeroize
+/// A key used for symmetric encryption.
+pub struct SecretKey([u8; 32]);
+
+impl SecretKey {
+    pub const SIZE: usize = 32;
+
+    pub fn new(key: [u8; Self::SIZE]) -> Self {
+        SecretKey::from(key)
+    }
+}
+
+impl From<[u8; SecretKey::SIZE]> for SecretKey {
+    fn from(secret: [u8; SecretKey::SIZE]) -> Self {
+        SecretKey(secret)
+    }
+}
+
+impl From<[u8; Nonce::SIZE]> for Nonce {
+    fn from(nonce: [u8; Nonce::SIZE]) -> Self {
+        Nonce(nonce)
+    }
+}
+
+impl From<SecretKey> for [u8; SecretKey::SIZE] {
+    fn from(SecretKey(bytes): SecretKey) -> Self {
+        bytes
+    }
+}
+
+impl From<Nonce> for [u8; Nonce::SIZE] {
+    fn from(Nonce(bytes): Nonce) -> Self {
+        bytes
+    }
+}
+
+type NonceSize = U12;
+type SecretKeySize = U32;
+
+/// Cryptographically seal and authenticate a message using the
+/// [ChaCha20-Poly1305][chacha20poly1305] algorithm for [AEAD][aead].
+/// Associated data is also authenticated, but not encrypted.
+///
+/// [aead]: https://en.wikipedia.org/wiki/Authenticated_encryption
+/// [chacha]: https://en.wikipedia.org/wiki/ChaCha20-Poly1305
+pub fn seal(
+    msg: &[u8],
+    key: SecretKey,
+    nonce: Nonce,
+    aad: &[u8],
+) -> Result<Box<[u8]>, ring_compat::aead::Error> {
+    let mut buffer = Vec::from(msg);
+    let aead = ChaCha20Poly1305::new(GenericArray::<u8, SecretKeySize>::from_slice(&<[u8;
+        SecretKey::SIZE]>::from(
+        key
+    )));
+    aead.encrypt_in_place(
+        GenericArray::<u8, NonceSize>::from_slice(&<[u8; Nonce::SIZE]>::from(nonce)),
+        aad,
+        &mut buffer,
+    )?;
+    Ok(Box::from(buffer))
+}
+
+pub type SecretBytes = Zeroizing<Box<[u8]>>;
+/// Cryptographically unseal and authenticate a message by supplying it
+/// alongside its associated data.
+pub fn open(
```

Comment:
should we not rather call this `seal`

Reply:
I assume you are referring to these:
- seal: `crypto_secretbox_easy`
- unseal: `crypto_secretbox_open_easy`

There is value in precedence of course, though I just looked at Python and Rust bindings, and one example of the former uses `encrypt` and `decrypt`, while using words `seal` and `unseal`, and one example of the latter just uses `encrypt` and `decrypt`.

I see wallet uses `unseal`, though it looks like a higher-level api: https://github.com/ntls-io/nautilus-wallet/blob/45db8a868fd5bf5cf3c03f65e9dd9b1fd43a9aeb/web-server/sgx-wallet-impl/src/schema/sealing.rs#L42.

I personally prefer using words that feel like they make more sense, especially to new users, even though a balance is needed to avoid confusing those familiar with the domain... my issue with `open` is that if feels too generic a word, and says nothing very little about the operation, which does more than just open data. I also feel that `seal` cannot be easily confused with `unseal`.

I don't know if that's compelling enough, so feel free to keep it as `open` if you still have strong feels.